crates/ethereum/src/lib.rs

```diff
@@ -255,4 +270,205 @@ impl EthereumApi for EthereumClient {
+
+        // Fetch the logs
+        let mut logs = Vec::new();
+        get_logs_recursive(
+            &provider,
+            &filter,
+            from_block.get(),
+            to_block.get(),
+            &mut logs,
+            10_000,
+        )
+        .await?;
+
+        tracing::debug!(
+            "Fetched {} `L1ToL2MessageLog` logs from {} to {}",
+            logs.len(),
+            from_block,
+            to_block
+        );
+
+        let logs: Vec<L1ToL2MessageLog> = logs
+            .into_iter()
+            .filter_map(|log| {
+                log.log_decode::<StarknetCoreContract::LogMessageToL2>()
+                    .ok()
+                    .map(|decoded| L1ToL2MessageLog {
+                        message_hash: H256::from(decoded.inner.message_hash().to_be_bytes()),
+                        l1_block_number: Some(log.block_number.expect("missing eth block number")),
+                        l1_tx_hash: log.transaction_hash.map(|hash| H256::from(hash.0)),
+                        l2_tx_hash: None,
+                    })
+            })
+            .collect();
+
+        Ok(logs)
+    }
+
+    /// Get state updates for a given address and block range
+    async fn get_state_updates(
+        &self,
+        address: &H160,
+        from_block: BlockNumber,
+        to_block: BlockNumber,
+    ) -> anyhow::Result<Vec<EthereumStateUpdate>> {
+        // Create a WebSocket connection
+        let ws = WsConnect::new(self.url.clone());
+        let provider = ProviderBuilder::new().on_ws(ws).await?;
+
+        // Create the StarknetCoreContract instance
+        let address = Address::new((*address).into());
+        let core_contract = StarknetCoreContract::new(address, provider.clone());
+
+        // Create the filter
+        let filter = core_contract.LogStateUpdate_filter().filter;
+
+        // Fetch the logs
+        let mut logs = Vec::new();
+        get_logs_recursive(
+            &provider,
+            &filter,
+            from_block.get(),
+            to_block.get(),
+            &mut logs,
+            2_000,
+        )
+        .await?;
+
+        let logs: Vec<EthereumStateUpdate> = logs
+            .into_iter()
+            .filter_map(|log| {
+                log.log_decode::<StarknetCoreContract::LogStateUpdate>()
+                    .ok()
+                    .map(|decoded| EthereumStateUpdate {
+                        state_root: get_state_root(decoded.inner.globalRoot),
+                        block_hash: get_block_hash(decoded.inner.blockHash),
+                        block_number: get_block_number(decoded.inner.blockNumber),
+                        l1_block_number: Some(BlockNumber::new_or_panic(
+                            log.block_number.expect("missing eth block number"),
+                        )),
+                    })
+            })
+            .collect();
+
+        Ok(logs)
+    }
+}
+
+/// Recursively fetches logs while respecting provider limits
+#[async_recursion]
```

Comment:
small nit: I think you can make do [without the special crate](https://github.com/eqlabs/pathfinder/commit/daea6b645620012864b710a8b5eeba57050b79d0)